Boot 64 bit using Limine

DrDeano opened the pull request:
Boot the kernel into 64 bit mode using the Limine bootloader.
Basic memory initialisation
Fixes arch agnostic tests
Part of #169

SamTebbs33 commented on build.zig:
```diff
@@ -64,9 +92,11 @@ pub fn build(b: *Builder) !void {
 
     const make_iso = switch (target.getCpuArch()) {
         .i386 => b.addSystemCommand(&[_][]const u8{ "./makeiso.sh", boot_path, modules_path, iso_dir_path, exec.getOutputPath(), ramdisk_path, output_iso }),
+        .x86_64 => b.addSystemCommand(&[_][]const u8{ "./makeiso_64.sh", b.exe_dir, exec.getOutputPath(), output_iso, ramdisk_path }),
```
Would it be feasible to have just one makiso.sh file and change behaviour based on the architecture being built?

DrDeano replied:
The two files don't have much in common, so is easier to have 2 separate bash files than switching inside the bash file for the build target. This might change later on though.

SamTebbs33 commented on src/kernel/arch/x86_64/arch.zig:
```diff
@@ -0,0 +1,458 @@
+    // Page directory
+    cr3: usize,
+    // Extra segments
+    gs: u32,
+    fs: u32,
+    es: u32,
+    ds: u32,
+
+    // Destination, source, base pointer
+    edi: u32,
+    esi: u32,
+    ebp: u32,
+    esp: u32,
+
+    // General registers
+    ebx: u32,
+    edx: u32,
+    ecx: u32,
+    eax: u32,
+
+    // Interrupt number and error code
+    int_num: u32,
+    error_code: u32,
+
+    // Instruction pointer, code segment and flags
+    eip: u32,
+    cs: u32,
+    eflags: u32,
+    user_esp: u32,
+    user_ss: u32,
```
Shouldn't most of these be u64 (usize)?

DrDeano replied:
Have removed this as will be different for 64 bit

SamTebbs33 commented on src/kernel/arch/x86_64/arch.zig:
```diff
@@ -0,0 +1,458 @@
+                .end = end,
+            });
+        }
+    }
```
If I understand correctly then `memory_size_bytes` will only be the amount of memory that isn't reserved. The memory profile is however expected to specify the total amount of memory, including what is reserved.

DrDeano replied:
I have reworked the memory size.

SamTebbs33 commented on build.zig:
```diff
@@ -36,7 +60,11 @@ pub fn build(b: *Builder) !void {
-    const main_src = "src/kernel/kmain.zig";
+    const main_src = switch (target.getCpuArch()) {
+        .i386 => "src/kernel/kmain.zig",
+        .x86_64 => "src/kernel/kmain_64.zig",
```
What I do in the aarch64 port is check for the architecture in kmain and skip certain parts of the initialisation if it's aarch64. You could consider doing that for x86_64 as well but either is fine as it's WIP.

DrDeano replied:
From offline discussion, until the 64 bit port is at the same level as the 32 bit port, these files will be merged as a lot of the kernel.zig code will need to be commented out for the 64 bit port to work. So creating a new kmain and using that will reduce the amount of commented code and make thing clearer.

SamTebbs33 reviewed (approved):
LGTM

DrDeano committed:
Boot 64 bit using Limine


Removed unused CpuState


Fixed memory size and alignments

Align the .text.boot section in the ELF to the page boundary as the KernelAndModules memory type assumes sections are aligned to 4096.

Any holes in the memory map provided by the bootloader are assumed unused and are reserved in the PMM.
Update to master Zig

Swap the startFn and context around.
